### DISCUSSION TITLE
Rendering: should Envelope, resampling, pan, and gain apply before or after stacked effects?

### DISCUSSION THREAD

**Discussion by Paul-Licameli** [audacity/audacity#3303] [Development]
Audacity's envelope, pan, and gain are in effect special-case, simple non-destructive effects that have long been built in.  They apply during playback and also during rendering of tracks for mix-and-render or for export.

With added effect stacks, this question arises; and whatever the decision, it must be done consistently in playback and in rendering, which do not use the same code path.

Pan and gain have always applied last, and I think that is uncontroversial.

Some decision is needed, whether envelope applies before or after effects.  It is less clear what is right.  What would the user expect?

Hypothetical example:
- generate a tone with amplitude 1
- add a hard limiter effect to the track stack
- add envelope control points that reduce amplitude

If envelope applies first, then the limiter may do nothing.  If after, then it is the first clipped sound that is then deamplified.

In favor of applying envelope first:  consider how tracks are displayed.  There has been no effort put into changing the display to match how the stacked effects will transform the sound.  But the envelope is indicated visually with a change in the displayed waveform.

If envelope is applied first in rendering, then at least the displayed waveform does match this intermediate step of the rendering calculation.  But if effects apply before the envelope, then the displayed waveform corresponds to no step of the rendering and is misleading.

Resampling: another matter, when the stored track sample rate does not match the output rate (for playback or for rendering to export).  As done now in realtime playback:  First envelope applies, then resampling, then effects, and only last, the gain and pan sliders.

If resampling is in the wrong place, there is work to do in the playack engine.  If not, then rendering should be kept consistent with playback.

**Comment by Paul-Licameli** [comment]
@SteveDaulton @petersampsonaudacity @Tantacrul @LWinterberg @jessjwilliamson

**Comment by jessjwilliamson** [comment]
I'll repeat here what I said on standup :)

Short answer - Before
Long answer - It's difficult as there is only 1 envelope and ideally it would be great to have automation on both pre and post RT effects. So in future we could make use of the Track panel gain slider as a post/after real time effect automation/envelope so that way user can fade out everything. For this case though it would make more sense having envelope as a pre RT effect processing.

**Reply by Tantacrul** [reply]
Agree. To do it afterwards would work against how it is generally done and would very possibly confuse users.

**Reply by Paul-Licameli** [reply]
Here is a related matter:  Resampling must apply, in case the sample rate of stored data and of the output rate (to speaker or to an export file).

I don't know if opinions are as strong about resampling, but it might have some influence on output of some effects that are implemented with algorithms like FFT.

As things work now in realtime playback, envelope applies first; then resampling; then stacked effects; finally gain and pan.

Should stacked effects really apply before the resampling?  (I hope not, it means more work otherwise)

If what realtime does now is right, then I must take care to implement rendering too consistently with it, so that what you hear really matches what you export.

**Reply by SteveDaulton** [reply]
> Short answer - Before

That's the opposite of how it works with destructive effects.
Personally I think it would be confusing if, for example, Audacity's "Distortion" effect worked differently when applied as a real-time effect to how it works as a destructive effect.

The obvious solution is to make the position of "envelope, pan, and gain" explicit within the interface. List them in real-time effects stack, either at the top, the bottom, or user defined.

**Reply by Paul-Licameli** [reply]
> > Short answer - Before
> 
> That's the opposite of how it works with destructive effects. Personally I think it would be confusing if, for example, Audacity's "Distortion" effect worked differently when applied as a real-time effect to how it works as a destructive effect.
> 
> The obvious solution is to make the position of "envelope, pan, and gain" explicit within the interface. List them in real-time effects stack, either at the top, the bottom, or user defined.

You are referring to the Envelope, and not the other matter of resampling that I mentioned.

I agree that making envelope, pan, and gain into things that you could insert where you will in the stack would be a very flexible user interface -- but it won't happen in 3.2 for sure.

**Reply by teetow** [reply]
From what I gather, most plugins expect to run at the project sample rate. I'm not sure what this means for Audacity (is there such a thing?), but the implication would be that whatever rate you're playing back at, is the rate at which the plugin should run. Any resampling necessary to facilitate that would then have to happen pre FX. Same thing during exporting, which would normally be the same sample rate.

It's probably worth pointing out that sample rates are normally completely transparent to the plugin user experience. I wouldn't expect a project set at 88200 Hz to suddenly show me EQ bands for 40 kHz, and indeed very few plugins I know of behave this way. I guess the assumption is that any plugin made for music production will be optimized for human hearing, regardless of project sample rate.

**Reply by Paul-Licameli** [reply]
Audacity has a “project” sample rate which mostly determines just the rate that new tracks use to store data.  Each track may have a different rate.

The rate  affects the frequency resolution available in Audacity’s spectrogram views.  Some plug-ins may also function as analyzers.  A noise reduction plug-in might display the spectrum of noise.

**Reply by crsib** [reply]
And there is a playback rate as well, which may be different that the project rate, From what I know - during the RT-stack processing we resample to the playback rate, not the project rate. This may or may not result in differences between the playback and rendering results.

**Reply by crsib** [reply]
And by the way - we can render into a different sample rate too, depending on what was requested

**Reply by SteveDaulton** [reply]
> I agree that making envelope, pan, and gain into things that you could insert where you will in the stack would be a very flexible user interface -- but it won't happen in 3.2 for sure.

I think that if it was indicated that the envelope "effect" came after other effects, then users are likely to assume that Gain and Pan do the same. With Gain, Pan and Envelope being applied after other effects, dynamic effects such as Compressor, Limiter, Gate... would behave the same whether applied as "real time" or "destructive" effects.

For 3.2, it may be sufficient to indicate this with a placeholder button:

![placeholder](https://user-images.githubusercontent.com/11681097/181606656-e0691f2b-a391-4cd4-b0e7-efdef147c90b.png)

**Comment by teetow** [comment]
This is why, in studio / DAW lingo, we make the distinction between "gain" (applied pre FX) and "volume" (applied post FX). You need both, but if I had to pick one, it'd be volume.

Its chief utility is just that it's post FX, and in essence "less destructive:" You'll use it to adjust the perceived loudness _without_ altering the incoming signal amplitude, and so the effects applied are kept intact. Applying it pre-FX will throw any kind of dynamic processing out of whack, and might leave the user frustrated why the noise gate suddenly start cutting out entire words.

That said, you already call it "gain," and seeing as it visually updates the waveform, it's then fair to say that Audacity only has pre-FX gain envelopes and not volume envelopes. My only hope is that you'll eventually have both.

**Reply by Paul-Licameli** [reply]
There are Audacity's envelope curves, but then there is also the other slider on each track that we (mis?-) call gain.

Envelope is now pre-FX and affects the display of the waveform.

The slider is post-FX and does not affect the display of the waveform.

**Reply by teetow** [reply]
Right, so renaming the slider to Volume (and, in the future, allowing Volume to be automated) would align with the common topology.

**Reply by jessjwilliamson** [reply]
If Audacity were using the today's DAW wordings, the slider on the track control panel that is currently called 'gain' would be renamed to 'volume'. And the envelope would be controlling the gain of the waveform. Maybe nothing to worry about right now but aware this can be confusing :)

**Comment by Paul-Licameli** [comment]
See issue #3305 for the good point @SteveDaulton raised

**Comment by Tantacrul** [comment]
As @teetow said, I think we'll need to expand the envelope tool at some point, so it can handle automate volume as well as gain. With the addition of RT effects, the need becomes a lot more obvious, since gain will change the desired sound as you increase / decrease it - especially with limiters, compressors, expanders, etc.
